idevice/src/services/misagent.rs

```diff
@@ -2,200 +2,316 @@
 //!
 //! Provides functionality for interacting with the misagent service on iOS devices,
 //! which manages provisioning profiles and certificates.
+//!
+//! Based on libimobiledevice implementation from SideStore
 
-use log::warn;
-use plist::Dictionary;
+use log::{debug, warn};
+use plist::{Dictionary, Value};
 
-use crate::{obf, Idevice, IdeviceError, IdeviceService, RsdService};
+use crate::{lockdown::LockdownClient, obf, Idevice, IdeviceError, IdeviceService};
 
-/// Client for interacting with the iOS misagent service
-///
-/// The misagent service handles:
-/// - Installation of provisioning profiles
-/// - Removal of provisioning profiles
-/// - Querying installed profiles
-pub struct MisagentClient {
-    /// The underlying device connection with established misagent service
-    pub idevice: Idevice,
+/// Error codes returned by misagent service
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MisagentError {
+    Success = 0,
+    InvalidArgument = -1,
+    PlistError = -2,
+    ConnectionFailed = -3,
+    RequestFailed = -4,
+    UnknownError = -256,
 }
 
-impl RsdService for MisagentClient {
-    fn rsd_service_name() -> std::borrow::Cow<'static, str> {
-        obf!("com.apple.misagent.shim.remote")
+impl From<IdeviceError> for MisagentError {
+    fn from(err: IdeviceError) -> Self {
+        match err {
+            IdeviceError::Plist(_) => MisagentError::PlistError,
+            IdeviceError::Socket(_) => MisagentError::ConnectionFailed,
+            _ => MisagentError::UnknownError,
+        }
     }
+}
 
-    async fn from_stream(stream: Box<dyn crate::ReadWrite>) -> Result<Self, IdeviceError> {
-        let mut stream = Idevice::new(stream, "");
-        stream.rsd_checkin().await?;
-        Ok(Self::new(stream))
+impl From<MisagentError> for IdeviceError {
+    fn from(err: MisagentError) -> Self {
+        match err {
+            MisagentError::Success => IdeviceError::UnexpectedResponse, // Shouldn't happen
+            MisagentError::InvalidArgument => IdeviceError::FfiInvalidArg,
+            MisagentError::PlistError => IdeviceError::UnexpectedResponse,
+            MisagentError::ConnectionFailed => IdeviceError::NoEstablishedConnection,
+            MisagentError::RequestFailed => IdeviceError::MisagentFailure,
+            MisagentError::UnknownError => IdeviceError::MisagentFailure,
+        }
     }
 }
 
+/// Client for interacting with the iOS misagent service
+///
+/// Handles provisioning profile installation, listing, and removal operations.
+/// Uses traditional lockdown service connection (not RSD) for maximum compatibility.
+pub struct MisagentClient {
+    /// The underlying device connection with established misagent service
+    pub idevice: Idevice,
+    /// Last error code from the service
+    pub last_error: i32,
```

Comment:
Why are we storing the last error? Ergonomic Rust should return the Error as a result, not store it in the structure.